File: ruby/004_CrossSum.rb
# 004 - Cross Sum（★2）
# 包除原理

H, W = gets.split.map(&:to_i)
A = Array.new(H) { gets.split.map(&:to_i) }

r_sum = A.map { _1.sum }
c_sum = A.transpose.map { _1.sum }

ans =
  H.times.map do |i|
    W.times.map do |j|
      r_sum[i] + c_sum[j] - A[i][j]
    end
  end

ans.each { puts _1.join(" ") }

File: kotlin/004_CrossSum.kt
/**
 * 004 - Cross Sum（★2）
 * 包除原理
 */

fun main() {
  val (H, W) = readLine()!!.split(" ").map(String::toInt)
  val A = List(H) { readLine()!!.split(" ").map(String::toInt) }

  val rSums = A.map { it.sum() }
  val cSums = (0 until W).map { c -> A.sumBy { it[c] } }

  for (i in 0 until H) {
    val ans = A[i].mapIndexed { j, v -> rSums[i] + cSums[j] - v }
    println(ans.joinToString(" "))
  }
}
